torchsde/_core/adjoint.py

```diff
@@ -299,24 +301,39 @@ def sdeint_adjoint(sde,
     return ys[0] if tensor_input else ys
 
 
-def _get_adjoint_sde(sde, params, logqp=False):
-    if sde.noise_type == NOISE_TYPES.diagonal:
-        if logqp:
-            return adjoint_sdes.AdjointSDEDiagonalLogqp(sde, params=params)
-        else:
-            return adjoint_sdes.AdjointSDEDiagonal(sde, params=params)
-
-    elif sde.noise_type == NOISE_TYPES.scalar:
-        if logqp:
-            return adjoint_sdes.AdjointSDEScalarLogqp(sde, params=params)
-        else:
-            return adjoint_sdes.AdjointSDEScalar(sde, params=params)
+def _check_and_select_default_adjoint_method(sde, adjoint_method: str) -> str:
+    sde_type, noise_type = sde.sde_type, sde.noise_type
 
-    elif sde.noise_type == NOISE_TYPES.additive:
-        if logqp:
-            return adjoint_sdes.AdjointSDEAdditiveLogqp(sde, params=params)
-        else:
-            return adjoint_sdes.AdjointSDEAdditive(sde, params=params)
+    if adjoint_method is None:  # Select the default based on noise type of forward.
```

Comment:
Currently unsupported combos for adjoint are

- ito
  - general: this is hard to make efficient and probably not very useful in practice. Doing this would take some time to derive a way to compute the correction terms fast. Frankly the correction terms are messy
  - scalar: grunt work
strat
  - additive: grunt work
  - scalar: grunt work
  - diagonal: Requires some thinking to make efficient

Apart from ito-general, the remaining would probably take me at least a month to complete and test if I work on it (apart from working on this, I also have other projects). 

On a broader note, I'm more interested in making these things useful than supplying features that might just not be very useful.

Reply:
Yeah, makes sense. Naive question: is it not possible to just define the appropriate f,g for the backward pass, and then feed it through the same machinery as on the forward pass?
I'm guessing that would work, but isn't done because it would be inefficient for some reason? It could be added as an easy fallback option so that the functionality is there.

Interesting regarding the logqp. I don't use it, but I think the idea is pretty neat. I can see that maintaining it is a big hassle, though.
Perhaps it could be put into a wrapper SDE that takes another SDE and augments the state: keep it around but don't try and do any special-casing for it.